lib/Runtime/Language/JavascriptConversion.cpp

```diff
@@ -301,25 +301,29 @@ namespace Js
             // For all other types, convert the key into a string and use that as the property name
             JavascriptString * propName = JavascriptConversion::ToString(key, scriptContext);
 
-            if (VirtualTableInfo<Js::PropertyString>::HasVirtualTable(propName))
+            PropertyString * propertyString = PropertyString::TryFromVar(propName);
+            if (propertyString == nullptr)
             {
-                PropertyString * propertyString = (PropertyString *)propName;
-                *propertyRecord = propertyString->GetPropertyRecord();
-            }
-            else if (VirtualTableInfo<Js::LiteralStringWithPropertyStringPtr>::HasVirtualTable(propName))
-            {
-                LiteralStringWithPropertyStringPtr * str = (LiteralStringWithPropertyStringPtr *)propName;
-                if (str->GetPropertyString())
-                {
-                    *propertyRecord = str->GetPropertyString()->GetPropertyRecord();
-                }
-                else
+                LiteralStringWithPropertyStringPtr * strWithPtr = LiteralStringWithPropertyStringPtr::TryFromVar(propName);
+                if (strWithPtr != nullptr)
                 {
-                    scriptContext->GetOrAddPropertyRecord(propName->GetString(), propName->GetLength(), propertyRecord);
-                    PropertyString * propStr = scriptContext->GetPropertyString((*propertyRecord)->GetPropertyId());
-                    str->SetPropertyString(propStr);
+                    if (!strWithPtr->GetPropertyString())
+                    {
+                        propertyString = strWithPtr->GetPropertyString();
+                    }
+                    else
+                    {
+                        scriptContext->GetOrAddPropertyRecord(propName->GetString(), propName->GetLength(), propertyRecord);
+                        propertyString = scriptContext->GetPropertyString((*propertyRecord)->GetPropertyId());
+                        strWithPtr->SetPropertyString(propertyString);
+                    }
                 }
             }
+
+            if (propertyString != nullptr)
```

Comment:
this seems to be changing the logic of this function beyond simple refactoring.
should this be an else-if instead of just if? I think that keeps the same behavior.

Reply:
as in, with the original if/else-if/else, only one block was executed. now, more than one can execute, logically. not sure if there is some property/invariant of the code itself that ensures the same outcome.

Reply:
semantically it's equivalent -- no matter what we are going to get the propertyRecord, but some of the paths have different perf characteristics. and you're right the new code isn't doing things quite right. i'll try to better organize this

---
In reply to: [141236387](https://github.com/Microsoft/ChakraCore/pull/3792#discussion_r141236387) [](ancestors = 141236387)

Reply:
thanks for the followup